firmware/EncoderPID2/EncoderPID2.ino

```diff
@@ -1,722 +1,784 @@
+
+    if (interrupt != NULL)
+    {
+        (*tim)->attachInterrupt(interrupt);
+    }
+}
+
+void initializeEncoder(Encoder *encoder, HardwareTimer **tim, PinName pin, void (*interrupt)(void))
+{
+    encoder->timer = *tim;
+    encoder->channel = STM_PIN_CHANNEL(pinmap_function(pin, PinMap_PWM));
+    encoder->pin = pin;
+    encoder->pulseCount = 0;
+    encoder->PIDvelocity = 0;
+
+    (*tim)->setMode(encoder->channel, TIMER_INPUT_CAPTURE_RISING, pin);
+
+    if (interrupt != NULL)
+    {
+        (*tim)->attachInterrupt(encoder->channel, interrupt);
+    }
+}
+
+void initializeMotor(Motor *motor, HardwareTimer **tim, PinName pin)
+{
+    motor->timer = *tim;
+    motor->channel = STM_PIN_CHANNEL(pinmap_function(pin, PinMap_PWM));
+    motor->pin = pin;
+    motor->pulseDuration = 1500;
+    motor->setpoint = 0.0;
+
+    (motor->timer)->setMode(motor->channel, TIMER_OUTPUT_COMPARE_PWM1, pin);
+    (motor->timer)->setCaptureCompare(motor->channel, motor->pulseDuration, MICROSEC_COMPARE_FORMAT);
+    // add pid func here?
+}
+
+// Manually set motor to desired speed
+void setMotorSpeed(Motor *motor, uint16_t newPulseDuration, uint8_t feedbackIndex)
+{
+    // char intbuf[20];
+    // sprintf(intbuf, "%d", newPulseDuration);
+    // nh.loginfo(intbuf);
+    if (newPulseDuration >= MIN_PULSE && newPulseDuration <= MAX_PULSE)
+    {
+        // motor->PIDduration = newPulseDuration;
+        motor->pulseDuration = (volatile uint16_t) newPulseDuration;
+        (motor->timer)->setCaptureCompare(motor->channel, newPulseDuration, MICROSEC_COMPARE_FORMAT);
+    }
+}
+
+// Set motor speed based off value in motor struct
+void updateMotorSpeed(Motor *motor)
+{
+    (motor->timer)->setCaptureCompare(motor->channel, motor->pulseDuration, MICROSEC_COMPARE_FORMAT);
+}
+
+// Set new desired velocity in motor struct
+void setSetpoint(Motor *motor, volatile double newSetpoint)
+{
+    motor->setpoint = newSetpoint;
+}
+
+uint16_t map_(const double val, const double in_min, const double in_max, const double out_min, const double out_max)
+{
+    // mitigate overflows by capping output values if input is out of range
+    if (val >= in_max)
+    {
+        return (uint16_t)out_max;
+    }
+    if (val <= in_min)
+    {
+        return (uint16_t)out_min;
+    }
+    return (uint16_t)round((val - in_min) * (out_max - out_min) / (in_max - in_min) + out_min);
+}
+
+// convert velocity between -4 and 4 to a value between 1000 and 2000
+uint16_t velocityToPulse(const double vel, const bool reverse = false)
+{
+    const double control_min = -4.0;
+    const double control_max = 4.0;
+
+    if (reverse)
+    {
+        return map_(vel, control_min, control_max, MIN_PULSE, MAX_PULSE);
+    }
+    return map_(vel, control_min, control_max, MAX_PULSE, MIN_PULSE);
+}
+
+// use calibration graphs to convert meters per second to desired PWM pulse for specific encoder
+uint16_t metersPerSecondToPulse(const double vel, const int encoderIndex)

```

Comment:
update this function to have a `reverse` parameter so that it works like `velocityToPulse`